source/tests/system/nifgen_driver_api_tests.cpp

```diff
@@ -0,0 +1,396 @@
+  }
+
+ private:
+  std::shared_ptr<::grpc::Channel> channel_;
+  std::unique_ptr<::nidevice_grpc::Session> driver_session_;
+  std::unique_ptr<fgen::NiFgen::Stub> nifgen_stub_;
+  std::unique_ptr<nidevice_grpc::SessionRepository> session_repository_;
+  std::unique_ptr<fgen::NiFgenLibrary> nifgen_library_;
+  std::unique_ptr<fgen::NiFgenService> nifgen_service_;
+  std::unique_ptr<::grpc::Server> server_;
+};
+
+TEST_F(nifgenDriverApiTest, PerformSelfTest_CompletesSuccessfuly)
+{
+  ::grpc::ClientContext context;
+  fgen::SelfTestRequest request;
+  request.mutable_vi()->set_id(GetSessionId());
+  fgen::SelfTestResponse response;
+  ::grpc::Status status = GetStub()->SelfTest(&context, request, &response);
+
+  EXPECT_TRUE(status.ok());
+  EXPECT_EQ(kfgenDriverApiSuccess, response.status());
+  EXPECT_EQ(0, response.self_test_result());
+  EXPECT_LT(0, response.self_test_message().size());
+}
+
+TEST_F(nifgenDriverApiTest, PerformReset_CompletesSuccessfuly)
+{
+  ::grpc::ClientContext context;
+  fgen::ResetRequest request;
+  request.mutable_vi()->set_id(GetSessionId());
+  fgen::ResetResponse response;
+  ::grpc::Status status = GetStub()->Reset(&context, request, &response);
+
+  EXPECT_TRUE(status.ok());
+  EXPECT_EQ(kfgenDriverApiSuccess, response.status());
+}
+
+TEST_F(nifgenDriverApiTest, SetAttributeViInt32_GetAttributeViInt32ReturnsSameValue)
+{
+  const char* channel_name = "";
+  const fgen::NiFgenAttributes attribute_to_set = fgen::NiFgenAttributes::NIFGEN_ATTRIBUTE_OUTPUT_MODE;
+  const ViInt32 expected_value = fgen::OutputMode::OUTPUT_MODE_NIFGEN_VAL_OUTPUT_ARB;
+  ::grpc::ClientContext context;
+  fgen::SetAttributeViInt32Request request;
+  request.mutable_vi()->set_id(GetSessionId());
+  request.set_channel_name(channel_name);
+  request.set_attribute_id(attribute_to_set);
+  request.set_attribute_value(expected_value);
+  fgen::SetAttributeViInt32Response response;
+  ::grpc::Status status = GetStub()->SetAttributeViInt32(&context, request, &response);
+
+  EXPECT_TRUE(status.ok());
+  EXPECT_EQ(kfgenDriverApiSuccess, response.status());
+  ViInt32 get_attribute_value = get_int32_attribute(channel_name, attribute_to_set);
+  EXPECT_EQ(expected_value, get_attribute_value);
+}
+
+TEST_F(nifgenDriverApiTest, SetAttributeViReal64_GetAttributeViReal64ReturnsSameValue)
+{
+  const char* channel_name = "";
+  const fgen::NiFgenAttributes attribute_to_set = fgen::NiFgenAttributes::NIFGEN_ATTRIBUTE_DIGITAL_GAIN;
+  const ViReal64 expected_value = -1.0;
+  ::grpc::ClientContext context;
+  fgen::SetAttributeViReal64Request request;
+  request.mutable_vi()->set_id(GetSessionId());
+  request.set_channel_name(channel_name);
+  request.set_attribute_id(attribute_to_set);
+  request.set_attribute_value(expected_value);
+  fgen::SetAttributeViReal64Response response;
+  ::grpc::Status status = GetStub()->SetAttributeViReal64(&context, request, &response);
+
+  EXPECT_EQ("", status.error_message());
```

Comment:
I think you can remove this check. I believe if the status is ok we can assume the `error_message` on the `status` is empty.

Reply:
Done.